controllers/k6_initialize.go

```diff
@@ -26,128 +26,55 @@ import (
-					"app":      "k6",
-					"k6_cr":    k6.Name,
-					"job-name": fmt.Sprintf("%s-initializer", k6.Name),
-				}),
-			}
-			podList = &corev1.PodList{}
-		)
-		if err := r.List(ctx, podList, listOpts); err != nil {
-			log.Error(err, "Could not list pods")
-			return false, err
-		}
-		if len(podList.Items) < 1 {
-			log.Info("No initializing pod found yet")
-			return false, nil
+	if k6.Status.Stage == "" {
+		log.Info("Changing stage of K6 status to initialization")
+		k6.Status.Stage = "initialization"
+		if err = r.Client.Status().Update(ctx, k6); err != nil {
+			log.Error(err, "Could not update status of custom resource")
+			return
 		}
 
-		// there should be only 1 initializer pod
-		if podList.Items[0].Status.Phase != "Succeeded" {
-			log.Info("Waiting for initializing pod to finish")
-			return false, nil
+		var initializer *batchv1.Job
+		if initializer, err = jobs.NewInitializerJob(k6, cli.ArchiveArgs); err != nil {
+			return res, err
 		}
 
-		// Here we need to get the output of the pod
-		// pods/log is not currently supported by controller-runtime client and it is officially
-		// recommended to use REST client instead:
-		// https://github.com/kubernetes-sigs/controller-runtime/issues/1229
-
-		config, err := rest.InClusterConfig()
-		if err != nil {
-			log.Error(err, "unable to fetch in-cluster REST config")
-			// don't return here
-			return false, nil
-		}
+		log.Info(fmt.Sprintf("Initializer job is ready to start with image `%s` and command `%s`",
+			initializer.Spec.Template.Spec.Containers[0].Image, initializer.Spec.Template.Spec.Containers[0].Command))
 
-		clientset, err := kubernetes.NewForConfig(config)
-		if err != nil {
-			log.Error(err, "unable to get access to clientset")
-			// don't return here
-			return false, nil
-		}
-		req := clientset.CoreV1().Pods(k6.Namespace).GetLogs(podList.Items[0].Name, &corev1.PodLogOptions{
-			Container: "k6",
-		})
-		ctx, cancel := context.WithTimeout(context.Background(), time.Second*60)
-		defer cancel()
-
-		podLogs, err := req.Stream(ctx)
-		if err != nil {
-			log.Error(err, "unable to stream logs from the pod")
-			// don't return here
-			return false, nil
+		if err = ctrl.SetControllerReference(k6, initializer, r.Scheme); err != nil {
+			log.Error(err, "Failed to set controller reference for the initialize job")
+			return
 		}
-		defer podLogs.Close()
 
-		buf := new(bytes.Buffer)
-		_, err = io.Copy(buf, podLogs)
-		if err != nil {
-			log.Error(err, "unable to copy logs from the pod")
-			return false, err
+		if err = r.Create(ctx, initializer); err != nil {
+			log.Error(err, "Failed to launch k6 test initializer")
+			return
 		}
+	}
 
-		if err := json.Unmarshal(buf.Bytes(), &inspectOutput); err != nil {
-			// this shouldn't normally happen but if it does, let's log output by default
-			log.Error(err, fmt.Sprintf("unable to marshal: `%s`", buf.String()))
-			return true, err
-		}
+	hasFinished, err := initJobHasFinalized(ctx, log, k6, r)
```

Comment:
The same as with `allK6RunnerPodsAreReadyToStart`: please keep the same structure as before. It has an added bonus of simplifying understanding of what changed and what didn't :slightly_smiling_face:

Reply:
Code removed